File: Java/libs/src/main/java/ninox360/util/HttpCommunicationUtil.java
package ninox360.util;

import okhttp3.*;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;

import javax.imageio.ImageIO;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Handles HTTPS setup, Packet Construction, Server interaction, Serialization, Deserialization,
 * Image requests, and Error Handling.
 */
public class HttpCommunicationUtil {
    static String ipAddressOut;
    static String ipBase = "https://localhost:3003";

    /**
     * Constructs a Nested Map representing a YAML packet and sets the routing address. Calls the serializer.
     *
     * @param serverSecret key which is provided by the remote control application on SI
     * @param clientSecret client secret given by the server
     * @param version      packet version
     * @param command      action which we are issuing
     * @param text         additional parameters for a command
     * @return constructed packet as a serialized String
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    public static String constructPacket(String serverSecret, String clientSecret, int version, String command, String text) throws IOException {
        Map<String, Object> packetName = new HashMap<>();
        Map<String, Object> innerPacket = new HashMap<>();

        try {
            if (clientSecret.isEmpty()) { //assume we are registering the device
                innerPacket.put("scanner_secret", serverSecret);
                innerPacket.put("version", version);

                packetName.put("PacketRemoteRegister", innerPacket);

                ipAddressOut = ipBase + "/register";
            } else {
                innerPacket.put("secret", clientSecret);
                innerPacket.put("value", text);
                innerPacket.put("command", command);

                packetName.put("PacketRemoteCommand", innerPacket);

                ipAddressOut = ipBase + "/command";
            }

            return serializeYaml(packetName);
        } catch (Exception e) {
           throw handleError(e, "Failed to create outbound packet prior to serialization");
        }
    }

    /**
     * Serialize/Format the yaml map into a string
     *
     * @param yamlPacket Nested Map which will be serialized
     * @return Formatted String
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private static String serializeYaml(Map<String, Object> yamlPacket) throws IOException {
        try {
            DumperOptions options = new DumperOptions();

            options.setIndent(2);
            options.setPrettyFlow(true);
            options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);

            Yaml yaml = new Yaml(options);

            return yaml.dump(yamlPacket);
        } catch (Exception e) {
            throw handleError(e, "Failed to serialize yaml packet");
        }
    }

    /**
     * Helper function to simplify request creation and execution
     *
     * @param packet which is being sent out
     * @param url    endpoint which the packet is being sent to
     * @return Response from the server
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private Response executeRequest(String packet, String url) throws IOException {
        OkHttpClient client = setupOkHttpClient();

        try {
            RequestBody body = RequestBody.create(packet, MediaType.parse("application/x-yaml"));

            Request request = new Request.Builder().url(url).post(body).build();

            return client.newCall(request).execute();
        } catch (Exception e) {
            throw handleError(e, "Error encountered during request execution");
        }
    }

    /**
     * Ingests a serialized YAML packet to send out to the server, manages the Yaml response
     *
     * @param packet Serialized YAML packet which is being sent out.
     * @param url,    URL endpoint.
     * @return Response Packet of the response contents for easy processing
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    public Map<String, Object> sendPacketAndDeserializeYaml(String packet, String url) throws IOException {
        // Execute the request and get the response
        try (Response response = executeRequest(packet, url)) {
            Objects.requireNonNull(response);
            if (!response.isSuccessful()) handleRejectedRequest(response);

            Objects.requireNonNull(response.body());
            return deserializeYamlPacket(response.body().string());
        } catch (Exception e) {
            throw handleError(e, "Failed to process command request");
        }
    }

    /**
     * Ingests a serialized YAML packet to send out to the server, manages the Image response
     *
     * @param packet, Serialized YAML packet which is being sent out.
     * @param url,    URL endpoint.
     * @return BufferedImage for further processing
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    public BufferedImage sendPacketAndDeserializeImage(String packet, String url) throws IOException {
        // Execute the request and get the response
        try (Response response = executeRequest(packet, url)) {
            Objects.requireNonNull(response);
            if (!response.isSuccessful()) handleRejectedRequest(response);

            Objects.requireNonNull(response.body());
            return getImageObject(response.body().bytes());
        } catch (Exception e) {
            throw handleError(e, "Failed to process image command request");
        }
    }

    /**
     * Extracts the yaml information and populates a global map for the response packet contents
     *
     * @param textIn raw string response which is being deserialized
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private static Map<String, Object> deserializeYamlPacket(String textIn) throws IOException {
        Yaml yaml = new Yaml();
        try {
            // load the full packet
            Map<String, Object> yamlResult = yaml.load(textIn);
            // extract the inner map and store it globally
            return ((Map<String, Object>) yamlResult.get("PacketRemoteResponse"));
        } catch (Exception e) {
            throw handleError(e, "Failed to Deserialize YAML Response Packet");
        }
    }

    /**
     * Processes the image bytes from the response and validates
     *
     * @param imageBytes response Bytes which we are saving locally
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private static BufferedImage getImageObject(byte[] imageBytes) throws IOException {
        // Convert the byte array to BufferedImage
        try (ByteArrayInputStream byteStream = new ByteArrayInputStream(imageBytes)) {
            BufferedImage image = ImageIO.read(byteStream);
            if (image == null) {
                throw new IOException("The byte array does not contain a valid image");
            }
            return image;
        } catch (IOException e) {
            throw handleError(e, "saveImage FAILED: - ImageIo.read failure, Image is not valid");
        }
    }

    /**
     * General Error handling, prints a specific message with the Exception.
     *
     * @param e   Exception to display.
     * @param msg Error message to provide additional context from the function.
     * @return IOException indicated an issue, printing the function specific message and exception details.
     */
    private static IOException handleError(Exception e, String msg) {
        System.err.println(msg + " Hit Exception: '" + e + "'");
        return new IOException(msg, e);
    }

    /**
     * Handle a failed or blocked response packet when trying to execute a request.
     *
     * @param response object returned back from the server.
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private void handleRejectedRequest(Response response) throws IOException {
        String responseBody = response.body() != null ? response.body().string() : "No response body";
        throw new IOException("Request failed: Status - '" + response.code() + "', Message - '" + response.message() + "' , Body - '" + responseBody + "'");
    }

    /**
     * SSL Management
     *
     * @return the modified client instance which we will communicate via
     * @throws IOException indicated an issue, printing the function specific message and exception details.
     */
    private static OkHttpClient setupOkHttpClient() throws IOException {
        try {
            // Create a trust manager that does not validate certificate chains
            final TrustManager[] trustAllCerts = new TrustManager[]{new X509TrustManager() {
                @Override
                public void checkClientTrusted(X509Certificate[] chain, String authType) {
                }

                @Override
                public void checkServerTrusted(X509Certificate[] chain, String authType) {
                }

                @Override
                public X509Certificate[] getAcceptedIssuers() {
                    return new X509Certificate[]{};
                }
            }};

            // Install the all-trusting trust manager
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new SecureRandom());

            // Create a ssl socket factory with our all-trusting manager
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            builder.sslSocketFactory(sslSocketFactory, (X509TrustManager) trustAllCerts[0]);
            builder.hostnameVerifier((hostname, session) -> true);

            return builder.build();
        } catch (Exception e) {
            throw handleError(e, "Failed to setup OkHttpClient");
        }
    }
}

File: Python/SIRemote/HttpCommunicationUtil.py
import requests
import yaml
from PIL import Image
from io import BytesIO
from typing import Dict, Any, Optional


def _get_image_object(image_data: bytes, ) -> Image:
    """
    Processes the image bytes from the response and validates

    :param image_data: byte sequence response which is being stored
    :return: PIL Image object for further manipulation
    :raises IOError: with the contextual message and exception for improved debugging
    """

    try:
        byte_stream = BytesIO(image_data)
        image = Image.open(byte_stream)

        # Verify that the image is valid by attempting to load it. This is not loaded in memory
        image.verify()

        # Reopen the image after verification to perform further operations
        byte_stream.seek(0)
        return Image.open(byte_stream)
    except Exception as e:
        raise handle_error(e, "saveImage FAILED: - Image failure, Image is not valid")


def _setup_http_client() -> requests.Session:
    """
    Helper function to create the HTTPS session, disabling any SSL verification and muting all warnings

    :return: session object
    :raises IOError: with the contextual message and exception for improved debugging
    """
    try:
        session = requests.Session()
        session.verify = False  # Disable SSL verification for all requests
        requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)
        return session
    except Exception as e:
        raise handle_error(e, "Problem encountered during HTTPS session creation")


def _execute_request(packet: str, url: str):
    """
    Executes an HTTP POST request with the given packet to the specified URL.

    :param packet: Serialized YAML packet to be sent.
    :param url: URL to send the request to.
    :return: HTTP response object.
    :raises IOError: with the contextual message and exception for improved debugging
    """
    try:
        client = _setup_http_client()
        headers = {'Content-Type': 'application/x-yaml'}
        return client.post(url, data=packet, headers=headers)
    except Exception as e:
        raise handle_error(e, "Error encountered during request execution")


def _serialize_yaml_packet(data: Dict[str, Any]) -> str:
    """
    Serializes a dictionary/Map to a YAML formatted string.

    :param data: Dictionary to be serialized.
    :return: YAML formatted string.
    :raises IOError: with the contextual message and exception for improved debugging
    """
    try:
        result: str = yaml.dump(data, default_flow_style=False)
        return result
    except Exception as e:
        raise handle_error(e, "Failed to serialize yaml packet")


def _deserialize_yaml_packet(yaml_data: str) -> Dict[str, Any]:
    """
    Deserializes a YAML formatted string to a dictionary/Map.

    :param yaml_data: YAML formatted string.
    :return: Deserialized dictionary.
    :raises IOError: with the contextual message and exception for improved debugging
    """
    try:
        outer_map: Dict[str, Any] = yaml.safe_load(yaml_data)
        return outer_map.get('PacketRemoteResponse', {})
    except Exception as e:
        raise handle_error(e, "Failed to Deserialize YAML Response Packet")


def handle_error(e: Exception, msg: str) -> IOError:
    """
    General Error handling, prints a specific message with the Exception.

    :param e: Exception to display.
    :param msg: Error message to provide additional context from the function.
    :return: IOError indicating an issue, printing the function specific message and exception details.
    :raises IOError: when an error occurs during the handling of the exception.
    """
    print(f"{msg} Hit Exception: '{e}'", file=sys.stderr)
    return IOError(f"{msg}: {str(e)}")


def handle_rejected_request(response: requests.Response) -> IOError:
    """
    Handle a failed or blocked response packet when trying to execute a request raising an exception if anything goes wrong.

    :param response: object returned back from the server.
    :return IOError: indicates an issue, printing the function specific message and exception details.
    """
    response_body = response.text if response.text else "No response body"
    return IOError(
        f"Request failed: Status - '{response.status_code}', Message - '{response.reason}' , Body - '{response_body}'")


class HttpCommunicationUtil:
    """
    Handles HTTPS setup, Packet Construction, Server interaction, Serialization, Deserialization,
    Image requests, and Error Handling
    """

    def __init__(self, ip_base='https://localhost:3003'):
        self.ipAddressOut = ""
        self.ipBase = ip_base

    def construct_packet(self, server_secret: str, client_secret: str, version: int, command: str, text: str) -> str:
        """
        Constructs a YAML packet based on the provided parameters and sets the target URL.

        :param server_secret: Secret key displayed on "Remote Control" page of the SI application.
        :param client_secret: Secret key returned from a registration packet.
        :param version: Version number.
        :param command: Command to be executed, if empty it is assumed a registration packet is requested.
        :param text: Text content included for some commands
        :return: Serialized YAML packet as a string.
        :raises IOError: with the contextual message and exception for improved debugging
        """
        try:
            if not command:
                # Create a registration packet if a command is not provided
                packet = {
                    'PacketRemoteRegister': {
                        'scanner_secret': server_secret,
                        'version': version
                    }
                }
                self.ipAddressOut = self.ipBase + '/register'
            else:
                # Create a command packet if a command is provided
                packet = {
                    'PacketRemoteCommand': {
                        'secret': client_secret,
                        'value': text,
                        'command': command
                    }
                }
                self.ipAddressOut = self.ipBase + '/command'

            # Serialize the YAML packet into a string and return it
            return _serialize_yaml_packet(packet)
        except Exception as e:
            raise handle_error(e, "Failed to create outbound packet prior to serialization")

    @classmethod
    def send_packet_and_deserialize_yaml(cls, packet: str, url: str) -> Optional[Dict[str, Any]]:
        """
        Sends a YAML packet to the specified URL and deserializes the YAML response. If a command request fails the error
         will be passed through otherwise an IOError will be raised.

        :param packet: Serialized YAML packet to be sent.
        :param url: URL to send the packet to.
        :return: Deserialized response dictionary, if available.
        :raises IOError: with the contextual message and exception for improved debugging
        """
        try:
            response = _execute_request(packet, url)

            if not response.ok:
                return handle_rejected_request(response)
            return _deserialize_yaml_packet(response.text)
        except Exception as e:
            raise handle_error(e, "Failed to process command request")

    @classmethod
    def send_packet_and_deserialize_image(cls, packet: str, url: str) -> Image:
        """
        Sends a YAML packet to the specified URL and returns the response as an image. If a command request fails the error
         will be passed through otherwise an IOError will be raised.

        :param packet: Serialized YAML packet to be sent.
        :param url: URL to send the packet to.
        :return: Image object from the response content.
        :raises IOError: with the contextual message and exception for improved debugging
        """
        try:
            response = _execute_request(packet, url)

            if not response.ok:
                raise handle_rejected_request(response)
            return _get_image_object(response.content)
        except Exception as e:
            raise handle_error(e, "Failed to process image command request")
